CS4001/4042 Assignment 1, Part B, Q1
---

Real world datasets often have a mix of numeric and categorical features – this dataset is one example. To build models on such data, categorical features have to be encoded or embedded.

PyTorch Tabular is a library that makes it very convenient to build neural networks for tabular data. It is built on top of PyTorch Lightning, which abstracts away boilerplate model training code and makes it easy to integrate other tools, e.g. TensorBoard for experiment tracking.

For questions B1 and B2, the following features should be used:   
- **Numeric / Continuous** features: dist_to_nearest_stn, dist_to_dhoby, degree_centrality, eigenvector_centrality, remaining_lease_years, floor_area_sqm
- **Categorical** features: month, town, flat_model_type, storey_range



---



In [1]:
!pip install pytorch_tabular[extra]

In [5]:
SEED = 42

import os

import random
random.seed(SEED)

import numpy as np
np.random.seed(SEED)

import pandas as pd

from pytorch_tabular import TabularModel
from pytorch_tabular.models import CategoryEmbeddingModelConfig
from pytorch_tabular.config import (
    DataConfig,
    OptimizerConfig,
    TrainerConfig,
)

> Divide the dataset (‘hdb_price_prediction.csv’) into train, validation and test sets by using entries from year 2019 and before as training data, year 2020 as validation data and year 2021 as test data.
**Do not** use data from year 2022 and year 2023.



In [6]:
# TODO: Enter your code here

df = pd.read_csv('hdb_price_prediction.csv')
df_train = df[df['year'] <= 2019]
df_val = df[df['year'] == 2020]
df_test = df[df['year'] == 2021]


> Refer to the documentation of **PyTorch Tabular** and perform the following tasks: https://pytorch-tabular.readthedocs.io/en/latest/#usage
- Use **[DataConfig](https://pytorch-tabular.readthedocs.io/en/latest/data/)** to define the target variable, as well as the names of the continuous and categorical variables.
- Use **[TrainerConfig](https://pytorch-tabular.readthedocs.io/en/latest/training/)** to automatically tune the learning rate. Set batch_size to be 1024 and set max_epoch as 50.
- Use **[CategoryEmbeddingModelConfig](https://pytorch-tabular.readthedocs.io/en/latest/models/#category-embedding-model)** to create a feedforward neural network with 1 hidden layer containing 50 neurons.
- Use **[OptimizerConfig](https://pytorch-tabular.readthedocs.io/en/latest/optimizer/)** to choose Adam optimiser. There is no need to set the learning rate (since it will be tuned automatically) nor scheduler.
- Use **[TabularModel](https://pytorch-tabular.readthedocs.io/en/latest/tabular_model/)** to initialise the model and put all the configs together.

In [91]:
# TODO: Enter your code here
num_col_names = ['dist_to_nearest_stn', 'dist_to_dhoby', 'degree_centrality', 'eigenvector_centrality', 'remaining_lease_years', 'floor_area_sqm']
cat_col_names = ['month', 'town', 'flat_model_type', 'storey_range']
data_config = DataConfig(
    target=[
        'resale_price'
    ],  # target should always be a list. Multi-targets are only supported for regression. Multi-Task Classification is not implemented
    continuous_cols=num_col_names,
    categorical_cols=cat_col_names,
)
trainer_config = TrainerConfig(
    auto_lr_find=True,  # Runs the LRFinder to automatically derive a learning rate
    batch_size=1024,
    max_epochs=50,
)
optimizer_config = OptimizerConfig('Adam')

model_config = CategoryEmbeddingModelConfig(
    task="regression",
    layers="50",  # Number of nodes in each layer
    activation="LeakyReLU",  # Activation between each layers
    learning_rate=1e-3,
)

tabular_model = TabularModel(
    data_config=data_config,
    model_config=model_config,
    optimizer_config=optimizer_config,
    trainer_config=trainer_config,
)
tabular_model.fit(train=df_train, validation=df_val)
result = tabular_model.evaluate(df_test)
pred_df = tabular_model.predict(df_test)
tabular_model.save_model("hdb_regression")
loaded_model = TabularModel.load_from_checkpoint("hdb_regression")

2023-10-02 11:14:55,597 - {pytorch_tabular.tabular_model:105} - INFO - Experiment Tracking is turned off


Global seed set to 42
2023-10-02 11:14:55,626 - {pytorch_tabular.tabular_model:473} - INFO - Preparing the DataLoaders
2023-10-02 11:14:55,635 - {pytorch_tabular.tabular_datamodule:290} - INFO - Setting up the datamodule for regression task
2023-10-02 11:14:55,700 - {pytorch_tabular.tabular_model:521} - INFO - Preparing the Model: CategoryEmbeddingModel
2023-10-02 11:14:55,740 - {pytorch_tabular.tabular_model:268} - INFO - Preparing the Trainer
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\trainer\connectors\accelerator_connector.py:589: LightningDeprecationWarning: The Trainer argument `auto_select_gpus` has been deprecated in v1.9.0 and will be removed in v2.0.0. Please use the function `pytorch_lightning.accelerators.find_usable_cuda_devices` instead.
  rank_zero_deprecation(
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
2023-10-02 11:14:55,781 - {p

Finding best initial lr:   0%|          | 0/100 [00:00<?, ?it/s]

c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: You called `self.log('train_loss', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: You called `self.log('train_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(
`Trainer.fit` stopped: `max_steps=100` reached.
Learning rate set to 0.5754399373371567
Restoring states from the checkpoint path at d:\micha\Michael\GitHub\Michael_Lee\CZ4042_lab\Assignment\.lr_find_7e711698-7b00-4b13-9f5f-34d200a6509e.ckpt
Restored all states from the checkpoint file at d:\micha\Michael\GitHub\Michael_Lee\CZ4042_lab\Assignment\.lr_find_7e711698-7b00-4b13-9f5f-34d200a6509e.ckpt
2023-10-02 11:14:59,354 - {pytorch_

┏━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃   ┃ Name             ┃ Type                      ┃ Params ┃
┡━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│ 0 │ _backbone        │ CategoryEmbeddingBackbone │  2.9 K │
│ 1 │ _embedding_layer │ Embedding1dLayer          │  1.5 K │
│ 2 │ head             │ LinearHead                │     51 │
│ 3 │ loss             │ MSELoss                   │      0 │
└───┴──────────────────┴───────────────────────────┴────────┘

Trainable params: 4.5 K                                                                                            
Non-trainable params: 0                                                                                            
Total params: 4.5 K                                                                                                
Total estimated model params size (MB): 0

Output()

2023-10-02 11:15:45,529 - {pytorch_tabular.tabular_model:584} - INFO - Training the model completed
2023-10-02 11:15:45,530 - {pytorch_tabular.tabular_model:1258} - INFO - Loading the best model


Output()

c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: 
You called `self.log('test_loss', ..., logger=True)` but have no logger configured. You can enable one by doing 
`Trainer(logger=ALogger(...))`
  rank_zero_warn(

c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: 
You called `self.log('test_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable 
one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(

c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\utilities\cloud_io.py:33: LightningDeprecationWarning: `pytorch_lightning.utilities.cloud_io.get_filesystem` has been deprecated in v1.8.0 and will be removed in v2.0.0. Please use `lightning_fabric.utilities.cloud_io.get_filesystem` instead.
  rank_zero_deprecation(
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\trainer\connectors\data_connector.py:224: PossibleUserWarning: The dataloader, test_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 8 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_loss         │       5050857984.0        │
│  test_mean_squared_error  │       5050857984.0        │
└───────────────────────────┴───────────────────────────┘

Output()

2023-10-02 11:15:47,946 - {pytorch_tabular.tabular_model:129} - INFO - Experiment Tracking is turned off
2023-10-02 11:15:47,946 - {pytorch_tabular.tabular_model:268} - INFO - Preparing the Trainer
c:\Users\micha\miniconda3\envs\cz4042_nn_alt\lib\site-packages\pytorch_lightning\trainer\connectors\accelerator_connector.py:589: LightningDeprecationWarning: The Trainer argument `auto_select_gpus` has been deprecated in v1.9.0 and will be removed in v2.0.0. Please use the function `pytorch_lightning.accelerators.find_usable_cuda_devices` instead.
  rank_zero_deprecation(
Trainer already configured with model summary callbacks: [<class 'pytorch_lightning.callbacks.rich_model_summary.RichModelSummary'>]. Skipping setting a default `ModelSummary` callback.
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs


> Report the test RMSE error and the test R2 value that you obtained.



In [94]:
from sklearn.metrics import r2_score, mean_squared_error
rsme = pow(mean_squared_error(pred_df['resale_price'], pred_df['resale_price_prediction']), 0.5)
r2_score = r2_score(pred_df['resale_price'], pred_df['resale_price_prediction'])

print("RSME Score:", rsme)
print("R2 Score:", r2_score)


RSME Score: 71069.38710449345
R2 Score: 0.8090553972101858


RSME Score: 71069

R2 Score: 0.81


> Print out the corresponding rows in the dataframe for the top 25 test samples with the largest errors. Identify a trend in these poor predictions and suggest a way to reduce these errors.



In [95]:
# TODO: Enter your code here
pred_df['pred_error'] = abs(pred_df['resale_price_prediction'] - pred_df['resale_price'])
pred_df = pred_df.sort_values('pred_error', ascending=False)
pred_df.head(25).reset_index()


index  month  year          town                full_address  \
0    92405     11  2021   BUKIT MERAH            46 SENG POH ROAD   
1    90957      6  2021   BUKIT BATOK  288A BUKIT BATOK STREET 25   
2   112128     12  2021      TAMPINES      156 TAMPINES STREET 12   
3    90608     12  2021        BISHAN       273B BISHAN STREET 24   
4    90521     10  2021        BISHAN        237 BISHAN STREET 22   
5   114254      9  2021     WOODLANDS      789 WOODLANDS AVENUE 6   
6    92442     11  2021   BUKIT MERAH          127D KIM TIAN ROAD   
7    98379     12  2021       HOUGANG        615 HOUGANG AVENUE 8   
8    92340     10  2021   BUKIT MERAH            56 HAVELOCK ROAD   
9    91871      6  2021   BUKIT MERAH         17 TIONG BAHRU ROAD   
10   90382      7  2021        BISHAN        251 BISHAN STREET 22   
11   90601     12  2021        BISHAN        251 BISHAN STREET 22   
12   93670     12  2021   BUKIT TIMAH              6 TOH YI DRIVE   
13   92504     12  2021   BUKIT MERAH            49 KIM PONG ROAD   
14   93825      8  2021  CENTRAL AREA       4 TANJONG PAGAR PLAZA   
15   90432      8  2021        BISHAN       275A BISHAN STREET 24   
16   92226      9  2021   BUKIT MERAH          96A HENDERSON ROAD   
17   92299     10  2021   BUKIT MERAH         36 MOH GUAN TERRACE   
18   88081      8  2021    ANG MO KIO    310A ANG MO KIO AVENUE 1   
19   92533     12  2021   BUKIT MERAH          2C BOON TIONG ROAD   
20  105695      6  2021    QUEENSTOWN              91 DAWSON ROAD   
21   98242     11  2021       HOUGANG   477B UPPER SERANGOON VIEW   
22   88421     12  2021    ANG MO KIO   104A ANG MO KIO STREET 11   
23  106192     12  2021    QUEENSTOWN              89 DAWSON ROAD   
24   90600     12  2021        BISHAN        251 BISHAN STREET 22   

      nearest_stn  dist_to_nearest_stn  dist_to_dhoby  degree_centrality  \
0     Tiong Bahru             0.581977       2.309477           0.016807   
1     Bukit Batok             1.292540      10.763777           0.016807   
2        Tampines             0.370873      12.479752           0.033613   
3          Bishan             0.776182       6.297489           0.033613   
4          Bishan             0.947205       6.663943           0.033613   
5       Woodlands             1.915461      16.660245           0.016807   
6     Tiong Bahru             0.686789       2.664024           0.016807   
7         Hougang             0.899849       8.828235           0.016807   
8     Tiong Bahru             0.451387       2.128424           0.016807   
9     Tiong Bahru             0.693391       2.058774           0.016807   
10     Ang Mo Kio             1.081018       6.939944           0.016807   
11     Ang Mo Kio             1.081018       6.939944           0.016807   
12   Beauty World             0.428356       8.948410           0.016807   
13    Tiong Bahru             0.468378       2.365532           0.016807   
14  Tanjong Pagar             0.451637       2.594828           0.016807   
15         Bishan             0.827889       6.370404           0.033613   
16    Tiong Bahru             0.586629       2.932814           0.016807   
17    Tiong Bahru             0.490926       2.278805           0.016807   
18     Ang Mo Kio             0.860056       7.263401           0.016807   
19    Tiong Bahru             0.656363       1.982722           0.016807   
20     Queenstown             0.745596       3.720593           0.016807   
21       Buangkok             1.296299      10.947164           0.016807   
22     Ang Mo Kio             1.216557       8.071776           0.016807   
23     Queenstown             0.658035       3.807573           0.016807   
24     Ang Mo Kio             1.081018       6.939944           0.016807   

    eigenvector_centrality                 flat_model_type  \
0                 0.047782                3 ROOM, Standard   
1                 0.000217            EXECUTIVE, Apartment   
2                 0.000229           EXECUTIVE,

In [98]:
good_pred_df = pred_df.sort_values('pred_error', ascending=True).reset_index()

print("Compare the cateogorical features for the 25 worst and 25 best predictions. The tables display the top 5 categories within each categorical feature for the 25 worst and 25 best predictions.")
for col in cat_col_names:
    print(f"===== Comparing for categorical category {col} =====")
    df_compare = pd.DataFrame()
    df_compare['Bad Category'] = pred_df[col].head(25).value_counts()[:5].index.tolist()
    df_compare['Bad Category Count'] = pred_df[col].head(25).value_counts()[:5].values.tolist()
    
    df_compare['Good Category'] = good_pred_df[col].head(25).value_counts()[:5].index.tolist()
    df_compare['Good Category Count'] = good_pred_df[col].head(25).value_counts()[:5].values.tolist()
    
    print(df_compare)

Compare the cateogorical features for the 25 worst and 25 best predictions. The tables display the top 5 categories within each categorical feature for the 25 worst and 25 best predictions.
===== Comparing for categorical category month =====
   Bad Category  Bad Category Count  Good Category  Good Category Count
0            12                  10              6                    4
1            11                   3              1                    4
2             6                   3              2                    3
3            10                   3              4                    3
4             8                   3              3                    3
===== Comparing for categorical category town =====
  Bad Category  Bad Category Count Good Category  Good Category Count
0  BUKIT MERAH                   8   BUKIT MERAH                    4
1       BISHAN                   6     TOA PAYOH                    2
2      HOUGANG                   2       GEYLANG               

In [99]:
pred_df[num_col_names].head(25).describe()

dist_to_nearest_stn  dist_to_dhoby  degree_centrality  \
count            25.000000      25.000000          25.000000   
mean              0.821817       6.118320           0.019496   
std               0.359394       3.861212           0.006288   
min               0.370873       1.982722           0.016807   
25%               0.581977       2.594828           0.016807   
50%               0.745596       6.370404           0.016807   
75%               1.081018       8.071776           0.016807   
max               1.915461      16.660245           0.033613   

       eigenvector_centrality  remaining_lease_years  floor_area_sqm  
count               25.000000              25.000000       25.000000  
mean                 0.023421              74.230000      118.680000  
std                  0.026168              15.730427       19.208765  
min                  0.000024              50.166667       88.000000  
25%                  0.006243              63.666667      113.000000  
50%                  0.008342              73.500000      120.000000  
75%                  0.047782              90.166667      126.000000  
max                  0.103876              96.750000      154.000000

In [100]:
good_pred_df[num_col_names].head(25).describe()

dist_to_nearest_stn  dist_to_dhoby  degree_centrality  \
count            25.000000      25.000000          25.000000   
mean              0.710519       8.340128           0.017479   
std               0.387744       4.525807           0.005381   
min               0.114837       0.687215           0.008403   
25%               0.439622       4.847078           0.016807   
50%               0.717746       8.162365           0.016807   
75%               0.923266      12.165743           0.016807   
max               1.787313      17.346043           0.033613   

       eigenvector_centrality  remaining_lease_years  floor_area_sqm  
count               25.000000              25.000000       25.000000  
mean                 0.018681              63.500000       89.200000  
std                  0.043785              12.934307       26.925824  
min                  0.000011              45.916667       45.000000  
25%                  0.000382              53.500000       65.000000  
50%                  0.004897              61.416667       91.000000  
75%                  0.018783              71.666667      110.000000  
max                  0.217454              92.083333      133.000000

The top 25 test samples with the largest errors, the following trends were observed.

- From the data analysis done, most of the errors in prediction seem to stem from the differences in the distrubution of the categorical features. 
- For the top 25 test samples with the largest errors, most of the 25 samples had month = 12. This could have contributed to the large errors observed.
- Other deviations for this top 25 test sample could include having town= 'BISHAN'.

Improvements to reduce errors.
- Normalisation of continous data can be performed. In the target label 'resale_price', the values are rather large compared to the data features the model is trained with. This could have contributed to the large errors observed. 
- Feature engineering can be done to both remove redundant features and also create features that will provide more useful information during training. 
- Different encoding techniques for categorical features can be studied. This can help the model learn and use the cateogorical features better.